aiohttp/client_middlewares.py

```diff
@@ -0,0 +1,58 @@
+"""Client middleware support."""
+
+from collections.abc import Awaitable, Callable
+
+from .client_reqrep import ClientRequest, ClientResponse
+
+__all__ = ("ClientMiddlewareType", "ClientHandlerType", "build_client_middlewares")
+
+# Type alias for client request handlers - functions that process requests and return responses
+ClientHandlerType = Callable[..., Awaitable[ClientResponse]]
+
+# Type for client middleware - similar to server but uses ClientRequest/ClientResponse
+ClientMiddlewareType = Callable[
+    [ClientRequest, ClientHandlerType], Awaitable[ClientResponse]
+]
+
+
+def build_client_middlewares(
+    handler: ClientHandlerType,
+    middlewares: tuple[ClientMiddlewareType, ...],
+) -> ClientHandlerType:
+    """
+    Apply middlewares to request handler.
+
+    The middlewares are applied in reverse order, so the first middleware
+    in the list wraps all subsequent middlewares and the handler.
+
+    This implementation avoids using partial/update_wrapper to minimize overhead
+    and doesn't cache to avoid holding references to stateful middleware.
+    """
+    if not middlewares:
+        return handler
+
+    # Optimize for single middleware case
```

Comment:
What's the difference in performance? Is it just due to how the variable is referenced?

Reply:
There is no function call to make it for the single handler case

Reply:
Oh, of course, I forgot we're rebuilding on every request.